[JavaScript] Logger for Telemetry

choupete opened the pull request:
Fixes #3058, implements dev spec in #3246 when combined with #3331

# Logging for Adaptive Cards

## Summary

This feature contains an extensible way to log events and messages to the console, the Microsoft 1DS telemetry platform, and any other logging/telemetry platform that implements the `IACProvider` interface. Logging is disabled by default. When enabled, log calls currently only occur when a card contains the `Input.Rating` element (see #3332).

![image](https://user-images.githubusercontent.com/42462906/62993862-1608b600-be0e-11e9-98ff-5aceb2a74738.png)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/3333)

dclaux commented on source/nodejs/adaptivecards/src/logging/IACProvider.ts:
```diff
@@ -0,0 +1,31 @@
+	The IACProvider interface. All provider implementations must follow this interface
+	in order to be used by the IACLogger.
+*/
+export interface IACProvider {
```
It is also debatable whether you need both a Logger and a Provider type. I think you could have a single Logger type, something along those lines:

```typescript
export enum LogLevel {
    Warning,
    Error,
    Verbose,
    Info,
    Event
}

export abstract class Logger {	
    abstract log(
        logLevel: LogLevel,
        event: string,
        eventSourceName: string,
        message: string,
        correlationID?: string,
        valueSet?: object);
}

export class LoggerCollection extends Logger {
    private _loggers: Logger[];

    log(
        logLevel: LogLevel,
        event: string,
        eventSourceName: string,
        message: string,
        correlationID?: string,
        valueSet?: object) {
        for (let logger of this._loggers) {
            logger.log(logLevel, event, eventSourceName, message, correlationID, valueSet);
        }
    }

    constructor(...loggers: Logger[]) {
        super();

        this._loggers = loggers;
    }
}

export class ConsoleLogger extends Logger {
    log(
        logLevel: LogLevel,
        event: string,
        eventSourceName: string,
        message: string,
        correlationID?: string,
        valueSet?: object) {
        // Emit even to console
    }
}

export class AdaptiveCard {
    ...
    static logger: Logger;
    ...
}

// Use a single ConsoleLogger
AdaptiveCard.logger = new ConsoleLogger();

// Use multiple loggers
AdaptiveCard.logger = new LoggerCollection(new ConsoleLogger(), new Ms1DSLogger());
```

shalinijoshi19 replied:
Thanks David! We are using a separate provider abstraction to isolate the transport from the core logging APIs. that way we dont have re-implement everything when switching providers/transports.

shalinijoshi19 replied:
wrt abstract/interface - indeed it is debatable and I'd like us to use interfaces with a default concrete implementation if that's ok. Will look at some of the gaps in your implemetnation though functionality/usability wise and get back.

dclaux replied:
My proposal achieves the exact same result (transport vs core logging)

I actually don't think that's ok for the reasons I mentioned. Please articulate the reasons you think it is better to stick with an interface and let's discuss further.

choupete replied:
Hi David,

In terms of why we decided to have both Logger and Provider classes:

- We expose an API that allows for logging different levels of messages in an intuitive way (logWarn, logError, etc.). However, this is abstracted by the ACLogger class and instead of several log methods a provider must implement, they are only required to log to two. Requiring these in a single Logger interface/abstract class would put unnecessary burden and redundancy on the developer implementing a custom provider.
- Separating ACLogger (what we implement) and ACProvider (what another party may implement) means that we can catch any errors from the ACProvider within the ACLogger class rather than catching them in the existing code.
- There are some parsing calls (e..g, the switch blocks in the renderCardTelemetry() method) that I am working to move out of card-elements.ts and into the ACLogger class. In the future when we add new events/card elements, we can use the ACLogger to carry out any intermediary steps between a log call and the sending of data.

dclaux replied:
Hey Peter, your second argument makes sense and is enough of a justification for having a logger + provider model, thanks for explaining.

That said, it is still debatable whether the provider model should be interface- or abstract class-based. I believe classes should always be used unless an interface is the only option (and in my opinion that happens only when two classes that do not/cannot extend the same base class need to be treated in a generic way).

In the current context, a class works just fine so that's what we should use IMO. Remember we can basically never change an interface (e.g. add members, especially methods) that's already shipped because it would break external code that implements that interface - in Windows, the overuse of interfaces has lead to ISomething followed by ISomething2 followed by ISomethingEx and all of that confusing nonsense. We can however add members to a class without breaking anything.

shalinijoshi19 replied:
Hey David, like I said interface vs abstract classes is going to be a debate that i deally dont want to get into. Apart from following the ISP, and fleshing out very specifically the Logger interface + provider interfaces resp, it also allows for easy mocking and testability. Also because of the fact that it's so concise, I dont really expect this interface to change much at all. And in teh case that it does the newer versions inmplement the new interface (because you can have multiple inheritance with interfaces) while the older ones continue to use the older interface and dont have access to the new functionality by design. What scenarios are you concerned will break ? Again, it's pretty standard practice to use and define simple interfaces with more elaborate concrete classes and versioning them esp for simple abstractions like logging-- There's existing solutions at play we've used as guidance: 
 https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger?view=aspnetcore-2.2 
https://github.com/Azure/azure-functions-host/wiki/ILogger

dclaux replied:
Abstract base classes also allow for easy mocking and testability. It's also very common practice to use abstract base classes over interfaces. It would be great if you could provide concrete arguments/examples that outline how using interfaces is actually better in this context, like I have done for classes. Saying that "you don't envision that the problems that definitely can occur when using interfaces will actually happen" doesn't make much sense to me; I'd rather choose an implementation with which the problems can't happen in the first place.

dclaux commented on source/nodejs/adaptivecards/src/logging/log-enums.ts:
```diff
@@ -0,0 +1,7 @@
+export enum LogLevel {
+	Warn,
+	Error,
+	Verbose,
```
What is verbose?

shalinijoshi19 replied:
This buckets operational logs that a component might want to trace to provide additinoal debugging context - any log for eg that isn't an operational error or a warning event classifies as verbose.

dclaux replied:
If any log that isn't Error or Warning classifies as Verbose, why are there also Info and Event? Are those nuances of Verbose? When is one supposed to use one over the other?

dclaux commented on source/nodejs/adaptivecards/src/logging/log-enums.ts:
```diff
@@ -0,0 +1,7 @@
+	Error,
+	Verbose,
+	Info,
+	Event
```
What is the difference between Info and Event?

shalinijoshi19 replied:
ERror/Verbose/Info are operational traces; Events are more schematized data representing feature 'metrics'

dclaux replied:
How do we actually use those differences when we pull data from the logs? Would anyone who's looking at a report from that data ever want to filter on verbose vs info vs event and what would they get out of it?

dclaux replied:
What is the guidance for third party developers who implement a custom element/action on which type of log to add?

shalinijoshi19 replied:
Same as they would do today in their regular development workflow: 
Think Info ==> additional context that a developer would like to have access to in case they have to debug something tricky; Doesnt get uploaded by default from the client machine to the servers. 
Warning ==> imp context when functionality is impaired potentially causing errors or compromising functionality downstack but not enough to abort operation/wrkflow at that point in the execution phase
Errors ==> logic errors which we've decided to abort processing for and flag back upto the caller.
Events ==> Events that help instrument pieces of the code in a bit to answer some questions identified upfront by the  developer/feature crew for the control

dclaux replied:
I get why we'd want Error, Warning and Info. I fail to see why we also need Verbose and Event. These might be useful when instrumenting and debugging your own code, but here the goal is for our code (e.g. the AC library) to log things in a way that can be useful for any developer. Where would we use the Verbose/Info/Event levels in the library?

shalinijoshi19 commented on source/nodejs/adaptivecards/src/logging/GUIDHelper.ts:
```diff
@@ -0,0 +1,62 @@
+const uuidv4 = require('uuid/v4');
```
Agreed on having a single story; I'd prefer a package being used in accordance with the language we are developing in vs handrolling our own - do we keep our implementation uptodate with the latest and greatest as far as the RFC is concerned? Is there a guideline that says we should handroll our own solutions vs rely on well-known community adopted and promoted library packages in TS?  Going forward that should be the approach regardless of language/platform we are developing in vs add the additional overhead of custom-rolling solutions

dclaux replied:
The generation of GUID isn't something that evolves. There is no security concern associated with it (especially in this context) and as far as I can tell the algorithm used by the library is the same as the one that's already in shared.ts.

But, the library is basically 10 times the size of what's in shared.ts (which is the only thing we need) and has itself a dependency on another library (crypto.)

We should stick to what's already in shared.ts. There is no reason to rely on an external library.

shalinijoshi19 replied:
Does every TS project need to handroll their own UUID implementations?

shalinijoshi19 replied:
why is it that we are OK pointing our customers to full fledged Markdown parsing libraries like Markedown-it for eg even though we support just a fraction of the MD processing it provides but are drawing the line here ourselves ? The general engineering guidance is not to reinvent the wheel when we can prevent it so am trying to understand this a bit more. Do you know of all the dependencies we have how many are pulling in that are adding to our bloat so we can prioritize and try to create a guideline for the future here?

dclaux replied:
The two (Markdown and UUID) really can't be compared...
- Markdown is not a simple thing to implement. For reference, here is the code from the React Native renderer: https://github.com/microsoft/AdaptiveCards/blob/master/source/community/reactnative/src/utils/markdown-formatter.js
- Markdown processing needs to be fast, reliable and secure. That quite an investment in terms of development cost and maintenance 
- Many web apps (MS Teams for instance) already use Markdown, via a third party library of their choice. They want to use that library with adaptivecards
- Teams that consume the TS renderer have made it clear multiple times that they care very much about the size of the adaptivecards library. We need to keep its size as low as possible and rely as little as possible (ideally not at all) on external libraries. In the context of Markdown, the code I pointed to above would add 13 KB to the library, which is a lot
- There is already an issue open to bake Markdown support into the library
- UUID generation, on the other hand, is simple, has no security concern associated with it and perf isn't really a concern.

To answer your last question, the adaptivecards package has 0 dependency. That is by design and I'd like to keep it that way unless using an external library is absolutely necessary, which isn't the case here.

shalinijoshi19 commented on source/nodejs/adaptivecards/src/card-elements.ts:
```diff
@@ -3836,6 +3840,32 @@ class ActionButton {
+	
+	render(alignment: Enums.ActionAlignment) {
+
+		var useGUID = GUIDHelper.getOrCreate().isGUIDtracked();
```
Thanks! Refer to https://github.com/airbnb/javascript as industry standard js guidelines per Windows guidance

shalinijoshi19 commented on source/nodejs/adaptivecards/src/logging/IACLogger.ts:
```diff
@@ -0,0 +1,68 @@
+/*
+	The IACLogger interface. All logger implementations in Adaptive Cards must follow this interface.
+*/
+export interface IACLogger {
```
I'd like to keep it for unit testing purposes

dclaux replied:
Can you clarify how it helps, and why it can't be declared in the unit tests themselves?

shalinijoshi19 replied:
Yeah  you are right and we should fix that up; We are missing an initializer/factory function. The idea is essentially, and i dont want to pull in the additional overhead of the ACLogger etc I just create a mock IACLogger that i can now inject into my tests and continue on my way testing just the control functionality.. similarly if i want to unit test the ACLogger implementation i can inject and mock up a mock IACProvider and not be dependent on any specific provider/transport

dclaux replied:
There *is* an initializer/factory function, it's the static getOrCreate method that lives off the ACLogger class. The implementation of that method always creates an instance of the ACLogger class, which makes the IACLogger interface pointless.

The only way for your tests to not "pull the overhead of the ACLogger class" would be for them to not load the adaptivecards library. IACLogger being part of that library, you will always have to load the library anyway, and thus you don't need IACLogger, you can use ACLogger directly.

shalinijoshi19 commented on source/nodejs/package.json:
```diff
@@ -37,5 +37,8 @@
 		"webpack-dev-server": "^3.7.2"
+	},
+	"dependencies": {
+		"uuid": "^3.3.2"
```
Other than the fact that we have something that may or may not a strict implementation of the UUID RFC is there a reason why we are averse to leveraging libraries that specialize and are widely adopted in teh community?

dclaux replied:
10 times the size while we use a 10th of it, dependency on yet another library, and achieves the exact same result.

shalinijoshi19 replied:
Interesting. What is the current guidance you are using on this project today?

dclaux replied:
The library doesn't have any external dependency and I would like to keep it that way unless using an external library is the only reasonable option, which isn't the case here. Please also see one of my previous comment re: Markdown vs UUID.

choupete commented on source/nodejs/adaptivecards/src/card-elements.ts:
```diff
@@ -6811,10 +6843,124 @@ export class AdaptiveCard extends ContainerWithActions {
+
+					switch (itemType) {
+						case "Input.ChoiceSet":
+							this.logSubItems(items[i], "choices", "Input.Choice");
```
Hi David,

Thanks for the feedback. We are logging Choices and Facts here in order to strictly keep track of how many elements are within a parent element (e.g., how many Choice elements are in a ChoiceSet). No other information is sent here.

choupete commented on source/nodejs/adaptivecards/src/card-elements.ts:
```diff
@@ -3836,6 +3840,32 @@ class ActionButton {
+			e.cancelBubble = true;
+			
+			if (this.action instanceof SubmitAction && useGUID) {
+				ACLogger.getOrCreate().logEvent("SubmitAction", "Action.Submit", guid);
```
- To address your first point, Action.Submit is only logged because that was the only action that was in scope for this iteration of this feature
- Regarding your fourth point, I agree that this would be a better alternative. I see that class "Action" has an "onExecute" property and I'd like to explore adding a hook here, although I am not seeing where this function is instantiated. Could you point me in the right direction for this?
- To address your fifth point, the GUID is used to correlate events on a unique card after that card is rendered. So if I render card 1 then render card 2, each would have a unique GUID. But later if I click a submit button belonging to card 1, then the hook will send a log containing the same GUID as when card 1 was rendered. I can do this at render time because I attach a hook to the onClick listener that holds even after the card is finished rendering.
- Addressing your final point, useGUID is true only if the card currently being rendered has an Input.Rating element on it. Using GUIDHelper to track this condition between the JSON parsing and the button render was used in order to avoid more encroachment into the existing code and avoid adding any global variables. There is a possibility that I could add more of this tracking logic within ACLogger to lessen the impact within existing code, and I plan on exploring this option.

Please let me know your thoughts. Thanks again for your feedback!

dclaux replied:
- Got it, that said I think it's a bit of a missed opportunity to address things in a generic way from the get go
- The code should probably live in the execute() method of the base Action class
- I am not sure I get it. I'll ping you offline to discuss
- Not sure this is the right approach. You seem to be using the term "useGUID" to mean something very specific, so maybe at least give it a more specific name?

choupete commented on source/nodejs/adaptivecards/src/card-elements.ts:
```diff
@@ -6811,10 +6843,124 @@ export class AdaptiveCard extends ContainerWithActions {
+		// ACLogger.getOrCreate().configureCustomProviders( { add providers here } );
+		ACLogger.getOrCreate().configureCustomProviders(new ConsoleProvider());
+
+		if (ACLogger.getOrCreate().isTelemetryEnabled()) {
```
Here isTelemetryEnabled() relies only on the condition that providers have been explicitly instantiated/not by the host. So this mirrors the process of checking if a logger has been set up or not.

dclaux replied:
Fair enough - although the implementation of isTelemetryEnabled only checks for the providers array to not be undefined, and not that it contains at least one provider.

But do we really want to have to sprinkle if (ACLogger.getOrCreate().isTelemetryEnabled()) everywhere? Can't we simply call ACLogger.getOrCreate().logXXX(...) and have that check whether telemetry is enabled?

And finally, we're using both "logging" and "telemetry" to mean the same thing here. We should remove the term "telemetry" and standardize on "log" everywhere.

choupete commented on source/nodejs/adaptivecards/src/card-elements.ts:
```diff
@@ -6811,10 +6843,124 @@ export class AdaptiveCard extends ContainerWithActions {
+	 * 
+	 * @param json - The JSON object that is being parsed
+	 */
+	private renderCardTelemetry(json: any): void {
```
Hi David,

I agree with your sentiment that another approach may be more ideal. I initially took this approach as to not add too many modifications to existing element classes as well as allow for easy tracking of only logging cards that have a rating input. However, reconsidering this, I see that while the former issue would still be valid, the impact could be minimized by adding a couple lines of code to each element’s internalRender() method to achieve the same effect. This would also ostensibly handle the Action edge cases as well as logging new elements. The latter issue is solved with the GUIDHelper keeping track of rating/no rating in parse().

I will explore this option of adding to internalRender() and update this commit with my findings. Please let me know what you think of this approach.

choupete replied:
Hi @dclaux ,

I prototyped a version in which logging calls are now made in each element's render() or internalRender() methods, adding just a couple lines of code at the beginning of each method to achieve this. This now solves your concerns of both edge cases and if/switch statements, as well as @shalinijoshi19 's concerns with extra bloat with the switch statements and conditional logic. Some things to consider if we go with this approach:

- Would it be more ideal if we created a telemetry wrapper class for each element class to house these new hooks? This would facilitate modularity but would add some extra bloat as well as add some extra dependencies (e.g. ColumnWithTelemetry would now extend ContainerWithTelemetry rather than Container). I am also unsure of how this would affect existing code - I can see that maybe I could add a new TypeRegistryWithTelemetry class and use this if telemetry metrics are desired, but I need to look more into this.
- One exception I found was in Facts due to the lack of explicit render() method. However, I see that this could be solved by logging within the parse() method of FactSet.

dclaux replied:
This sounds like it's going in the right direction, but I would have to see your code to have a better idea.

dclaux commented on source/nodejs/adaptivecards/src/card-elements.ts:
```diff
@@ -3841,21 +3841,19 @@ class ActionButton {
-		var useGUID = GUIDHelper.getOrCreate().isGUIDtracked();
-		var guid = GUIDHelper.getOrCreate().getGUID().toString();
+    render(alignment: Enums.ActionAlignment) {
+        let useGUID = GUIDHelper.getOrCreate().isGUIDtracked();
```
More on this - applications usually render multiple different cards. All these cards will share the same GUID given it's global. That doesn't seem like that's what you're trying to accomplish.

choupete commented:
@dclaux @shalinijoshi19 

In these most recent commits, I:

- Added a static logger variable to the AdaptiveCard class in order to allow for IACLogger extensibility (a user is required to only change one line if they'd like to reimplement IACLogger) and reduce the number of calls to getOrCreate().

- Refactored logging in order to include logging calls in:

1) the render() method of the CardElement class
2) the render() method and execute() methods of the Action class

This solves the issue of logging edge cases, as well as removes the added bloat of if/else and switch case statements in card-elements.ts. However, with this new approach comes some new caveats:

- a new private guid property was added to the Action class. This was required in order to have the correlationID be attached to the card instance, allowing the correlation remain the same between render and submit click events.
- the valueSet is no longer populated for Input.Rating telemetry metrics. This is because there is no trivial way to retrieve information such as the maxValue field and the iconSelected field from the parent CardElement class. One solution would be to add a valueSet property of type object in the CardElement class that all children classes can inherit. It would then be up to the individual element to decide which information to send. Thus, CardElement would be able to reference this property and retrieve the value at render time.

Please let me know your thoughts. Thanks!

paulcam206 commented:
closing out old PRs - feel free to revive this PR and resubmit for review
